src/main/java/org/testng/ReporterConfig.java

```diff
@@ -94,10 +97,18 @@ public IReporter newReporterInstance() {
     if (!(tmp instanceof IReporter)) {
       throw new TestNGException(m_className + " is not a IReporter");
     }
-
     IReporter result = (IReporter) tmp;
+
+    Object config;
```

Comment:
We could have done 

```java
Object config = result;
if (result instanceof IConfiguredReporter) {
       config = ((IConfiguredReporter) result).getConfig();
} else {
    LOGGER.warn(config.getClass().getName() + " does not implement " + IConfiguredReporter.class.getName() + ".  TestNG may not be able to alter the properties of this reporter in runtime");
}
```

Reply:
The warning is not relevant here because the default kind of configured reporter is the POJO way.

Reply:
@juherr - I agree. But then how are we going to inform users who are trying to access the configuration via a config object but its missing  (Which is the use case that this PR is trying to address). We need to have a way to inform the users that they are missing something which could lead to undesirable behaviors as well right ?

Reply:
You mean if a getConfig is available without implementing the interface? I suppose it will fail later.
Maybe we can add a fallback with a warning.

Reply:
@juherr - Yes. Now with this changeset, the getConfig would be invoked only if the interface type matches. But it won't match because `instanceof` will fail. The code will try to retrieve the parameters from the reporter itself, instead of retrieving them from the reporter config.

Reply:
There's a minimally used portion of the JavaBeans API that supports nested beans. This is the BeanContext package. I'm seeing if I can get the nested bean properties to be accessible with a few additional revisions.

Reply:
XMLReporter will extend BeanContextSupport, and XMLReporterConfig will extend BeanContextChildSupport.

Reply:
The context API is way more involved than the scenario calls for. I think what we have here is the right approach.

Reply:
@sbabcoc - With the new refactored changes, we don't seem to be handing the below condition with a proper error message (which is what prompted you to raise this PR)

User is expecting to set a configuration property that is exposed via the `ReporterConfig` object which is obtained via `getConfig()` method, but the `IReporter` implementation that he/she is working with doesn't implement `IConfiguredReporter` interface and so the implementation tries to query it via the regular JavaBean convention. How do we warn those users of the mistake that they are making?

The other observation that I had was that, we can now get rid of the `else` block by doing something like this

```java
Object config = result;
if (result instanceof IConfiguredReporter) {
       config = ((IConfiguredReporter) result).getConfig();
} 
```

Reply:
My original implementation would have handled this, but the implementing "by convention" to discover the `getConfig()` method via reflection was too freeform. I think we should either use this "configured reporter" approach, with the nested bean published by the `getConfig()` method of the interface, or go whole-hog to support nested beans via the bean context API. In my mind, this second option is way too complicated for the minimal configurability required by the typical reporter.
We could add some details to the warning, though:
* Make sure you've got the correct spelling for your reporter property name.
* If your reporter uses a nested configuration object, implement the `IConfiguredReporter` interface.